Faros interceptor

ilagunap opened the pull request:
First version of the FAROS interceptor:
- Code placed in interception/ directory
- Technique uses LD_PRELOAD to intercept clang, clang++, gcc, and g++
- It add the corresponding flags automatically
- To use, run:  "faros make"
- Some tests are added in test directory

ilagunap committed:
adding what FAROS is and fix a typo

ilagunap committed:
Merge branch 'develop' of github.com:LLNL/FAROS into develop

ilagunap committed:
Added first version of FAROS interceptor

ggeorgakoudis commented on interception/clang_faros.py:
```diff
@@ -0,0 +1,76 @@
+
+  # Is the command a link command?
+  def isLinkCommand(self) -> bool:
+    if ('-c' not in self.parameters and 
```
Let's remove the isLinkComand() check since clang can compile and link with a single invocation that we want to intercept too. Adding the "-fsave-optimization-record" flag, even if only linking, should not cause issues.

ggeorgakoudis commented on interception/faros:
```diff
@@ -0,0 +1 @@
+faros.py
```
Move to the root path of the repo, will simplify importing libs

ggeorgakoudis commented on interception/faros.py:
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env python3
```
Move to the root path of the repo, will simplify importing libs

ggeorgakoudis commented on interception/exceptions.py:
```diff
@@ -0,0 +1,14 @@
+
```
Clean up whitespace

ggeorgakoudis commented on interception/tests/test_mpi_openmp/Makefile:
```diff
@@ -0,0 +1,11 @@
+
```
Clean whitespace

ggeorgakoudis commented on interception/tests/test_mpi_openmp/main.c:
```diff
@@ -0,0 +1,28 @@
+
```
Clean whitespace

ggeorgakoudis commented on interception/tests/test_mpi_openmp/test_mpi_openmp.py:
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env python
```
Move tests under root path

ggeorgakoudis commented on interception/tests/test_simple_openmp/Makefile:
```diff
@@ -0,0 +1,11 @@
+
```
Clean whitespace

ggeorgakoudis commented on interception/tests/test_simple_openmp/main.c:
```diff
@@ -0,0 +1,23 @@
+
```
Clean whitespace

ggeorgakoudis reviewed (changes requested):
Looks good. Minor changes:

- Remove intercept.so
- Remove isLinkCommand code and check
- Move faros.py and symlink under the root path
- Add comments, license header as appropriate
- Clean up whitespace

ilagunap committed:
Merge branch 'LLNL:develop' into develop

ilagunap committed:
Added support for CMake files, fixed bugs, added more tests

ilagunap commented:
I would like to leave the interceptor code for some time in a separate directory until it's more stable.
I added:
1. Support for CMake files
2. A few more tests
3. Fixed a few bugs
4. Cleaned up intercept.c

ggeorgakoudis reviewed (changes requested):
Looks good, two things left:

- Remove the binary intercept.so
- Remove any necessary remaining whitespace

ilagunap committed:
removed unneeded whitespace

ilagunap committed:
adding .so to gitignore
